// aws-cdk-lib/aws-stepfunctions/lib/task-credentials.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.TaskRole = void 0;
const jsiiDeprecationWarnings = require("../../.warnings.jsii.js");
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const fields_1 = require("./fields");
/**
 * Role to be assumed by the State Machine's execution role for invoking a task's resource.
 *
 * @see https://docs.aws.amazon.com/step-functions/latest/dg/concepts-access-cross-acct-resources.html
 * @see https://docs.aws.amazon.com/step-functions/latest/dg/amazon-states-language-task-state.html#task-state-fields
 */
class TaskRole {
    /**
     * Construct a task role retrieved from task inputs using a json expression
     *
     * @param expression json expression to roleArn
     *
     * @example
     *
     * sfn.TaskRole.fromRoleArnJsonPath('$.RoleArn');
     */
    static fromRoleArnJsonPath(expression) {
        return new JsonExpressionTaskRole(expression);
    }
    /**
     * Construct a task role based on the provided IAM Role
     *
     * @param role IAM Role
     */
    static fromRole(role) {
        try {
            jsiiDeprecationWarnings.aws_cdk_lib_aws_iam_IRole(role);
        }
        catch (error) {
            if (process.env.JSII_DEBUG !== "1" && error.name === "DeprecationError") {
                Error.captureStackTrace(error, this.fromRole);
            }
            throw error;
        }
        return new IamRoleTaskRole(role);
    }
}
_a = JSII_RTTI_SYMBOL_1;
TaskRole[_a] = { fqn: "aws-cdk-lib.aws_stepfunctions.TaskRole", version: "2.76.0" };
exports.TaskRole = TaskRole;
class JsonExpressionTaskRole extends TaskRole {
    constructor(expression) {
        super();
        this.roleArn = fields_1.JsonPath.stringAt(expression);
        this.resource = '*';
    }
}
class IamRoleTaskRole extends TaskRole {
    constructor(role) {
        super();
        this.roleArn = role.roleArn;
        this.resource = role.roleArn;
    }
}
