// aws-cdk-lib/aws-ec2/lib/machine-image/amazon-linux-2022.js
"use strict";
var _a, _b;
Object.defineProperty(exports, "__esModule", { value: true });
exports.AmazonLinux2022ImageSsmParameter = exports.AmazonLinux2022Kernel = void 0;
const jsiiDeprecationWarnings = require("../../../.warnings.jsii.js");
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const common_1 = require("./common");
/**
 * Amazon Linux 2022 kernel versions
 */
class AmazonLinux2022Kernel {
    constructor(version) {
        this.version = version;
    }
    /**
     * Generate a string representation of the kernel
     */
    toString() {
        return `kernel-${this.version}`;
    }
}
_a = JSII_RTTI_SYMBOL_1;
AmazonLinux2022Kernel[_a] = { fqn: "aws-cdk-lib.aws_ec2.AmazonLinux2022Kernel", version: "2.76.0" };
/**
 * The latest kernel version currently available in a published AMI.
 *
 * When a new kernel version is available for an al2022 AMI this will be
 * updated to contain the latest kernel version and will cause your instances
 * to be replaced. Do not store stateful information on the instance if you are
 * using this version.
 */
AmazonLinux2022Kernel.CDK_LATEST = new AmazonLinux2022Kernel('5.15');
/**
 * The default kernel version for Amazon Linux 2022 is 5.15 and
 * the SSM parameter does not include it in the name
 * (i.e. /aws/service/ami-amazon-linux-latest/amzn2022-ami-kernel-default-x86_64)
 */
AmazonLinux2022Kernel.DEFAULT = new AmazonLinux2022Kernel('default');
/**
 * Kernel version 5.15
 */
AmazonLinux2022Kernel.KERNEL_5_15 = new AmazonLinux2022Kernel('5.15');
exports.AmazonLinux2022Kernel = AmazonLinux2022Kernel;
/**
 * A SSM Parameter that contains the AMI ID for Amazon Linux 2023
 */
class AmazonLinux2022ImageSsmParameter extends common_1.AmazonLinuxImageSsmParameterBase {
    /**
     * Generates a SSM Parameter name for a specific amazon linux 2022 AMI
     *
     * Example values:
     *
     *     "/aws/service/ami-amazon-linux-latest/al2022-ami-kernel-5.15-x86_64",
     *     "/aws/service/ami-amazon-linux-latest/al2022-ami-kernel-default-x86_64",
     *     "/aws/service/ami-amazon-linux-latest/al2022-ami-minimal-kernel-5.15-arm64",
     *     "/aws/service/ami-amazon-linux-latest/al2022-ami-minimal-kernel-5.15-x86_64",
     *     "/aws/service/ami-amazon-linux-latest/al2022-ami-kernel-5.15-arm64",
     *     "/aws/service/ami-amazon-linux-latest/al2022-ami-minimal-kernel-default-arm64",
     *     "/aws/service/ami-amazon-linux-latest/al2022-ami-minimal-kernel-default-x86_64",
     *     "/aws/service/ami-amazon-linux-latest/al2022-ami-kernel-default-arm64",
     */
    static ssmParameterName(props) {
        try {
            jsiiDeprecationWarnings.aws_cdk_lib_aws_ec2_AmazonLinux2022ImageSsmParameterProps(props);
        }
        catch (error) {
            if (process.env.JSII_DEBUG !== "1" && error.name === "DeprecationError") {
                Error.captureStackTrace(error, this.ssmParameterName);
            }
            throw error;
        }
        const edition = (props && props.edition) || common_1.AmazonLinuxEdition.STANDARD;
        const parts = [
            common_1.AmazonLinuxGeneration.AMAZON_LINUX_2022,
            'ami',
            edition !== common_1.AmazonLinuxEdition.STANDARD ? edition : undefined,
            props.kernel?.toString(),
            props.cpuType ?? common_1.AmazonLinuxCpuType.X86_64,
        ].filter(x => !!x);
        return '/aws/service/ami-amazon-linux-latest/' + parts.join('-');
    }
    constructor(props) {
        super({
            parameterName: AmazonLinux2022ImageSsmParameter.ssmParameterName(props),
            cachedInContext: props.cachedInContext,
            userData: props.userData,
        });
        try {
            jsiiDeprecationWarnings.aws_cdk_lib_aws_ec2_AmazonLinux2022ImageSsmParameterProps(props);
        }
        catch (error) {
            if (process.env.JSII_DEBUG !== "1" && error.name === "DeprecationError") {
                Error.captureStackTrace(error, AmazonLinux2022ImageSsmParameter);
            }
            throw error;
        }
    }
}
_b = JSII_RTTI_SYMBOL_1;
AmazonLinux2022ImageSsmParameter[_b] = { fqn: "aws-cdk-lib.aws_ec2.AmazonLinux2022ImageSsmParameter", version: "2.76.0" };
exports.AmazonLinux2022ImageSsmParameter = AmazonLinux2022ImageSsmParameter;
