// aws-cdk-lib/aws-ec2/lib/machine-image/amazon-linux2.js
"use strict";
var _a, _b;
Object.defineProperty(exports, "__esModule", { value: true });
exports.AmazonLinux2ImageSsmParameter = exports.AmazonLinux2Kernel = void 0;
const jsiiDeprecationWarnings = require("../../../.warnings.jsii.js");
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const common_1 = require("./common");
/**
 * Amazon Linux 2 kernel versions
 */
class AmazonLinux2Kernel {
    constructor(version) {
        this.version = version;
    }
    /**
     * Generate a string representation of the kernel
     */
    toString() {
        return this.version === 'default'
            ? undefined
            : `kernel-${this.version}`;
    }
}
_a = JSII_RTTI_SYMBOL_1;
AmazonLinux2Kernel[_a] = { fqn: "aws-cdk-lib.aws_ec2.AmazonLinux2Kernel", version: "2.76.0" };
/**
 * The latest kernel version currently available in a published AMI.
 *
 * When a new kernel version is available for an amzn2 AMI this will be
 * updated to contain the latest kernel version and will cause your instances
 * to be replaced. Do not store stateful information on the instance if you are
 * using this version.
 */
AmazonLinux2Kernel.CDK_LATEST = new AmazonLinux2Kernel('5.10');
/**
 * The default kernel version for Amazon Linux 2 is 4.14 and
 * Linux 4.14.311-233.529.amzn2.x86_64
 * the SSM parameter does not include it in the name
 * (i.e. /aws/service/ami-amazon-linux-latest/amzn2-ami-hvm-x86_64-gp2)
 */
AmazonLinux2Kernel.DEFAULT = new AmazonLinux2Kernel('default');
/**
 * Kernel version 5.10
 */
AmazonLinux2Kernel.KERNEL_5_10 = new AmazonLinux2Kernel('5.10');
exports.AmazonLinux2Kernel = AmazonLinux2Kernel;
/**
 * A SSM Parameter that contains the AMI ID for Amazon Linux 2
 */
class AmazonLinux2ImageSsmParameter extends common_1.AmazonLinuxImageSsmParameterBase {
    /**
     * Generates a SSM Parameter name for a specific amazon linux 2 AMI
     *
     * Example values:
     *
     *     "/aws/service/ami-amazon-linux-latest/amzn2-ami-hvm-x86_64-ebs",
     *     "/aws/service/ami-amazon-linux-latest/amzn2-ami-hvm-x86_64-gp2",
     *     "/aws/service/ami-amazon-linux-latest/amzn2-ami-kernel-5.10-hvm-x86_64-ebs",
     *     "/aws/service/ami-amazon-linux-latest/amzn2-ami-hvm-arm64-gp2",
     *     "/aws/service/ami-amazon-linux-latest/amzn2-ami-minimal-hvm-arm64-ebs",
     *     "/aws/service/ami-amazon-linux-latest/amzn2-ami-kernel-5.10-hvm-arm64-gp2",
     *     "/aws/service/ami-amazon-linux-latest/amzn2-ami-kernel-5.10-hvm-x86_64-gp2",
     *     "/aws/service/ami-amazon-linux-latest/amzn2-ami-minimal-hvm-x86_64-ebs"
     */
    static ssmParameterName(props) {
        try {
            jsiiDeprecationWarnings.aws_cdk_lib_aws_ec2_AmazonLinux2ImageSsmParameterProps(props);
        }
        catch (error) {
            if (process.env.JSII_DEBUG !== "1" && error.name === "DeprecationError") {
                Error.captureStackTrace(error, this.ssmParameterName);
            }
            throw error;
        }
        const edition = (props && props.edition) || common_1.AmazonLinuxEdition.STANDARD;
        const parts = [
            common_1.AmazonLinuxGeneration.AMAZON_LINUX_2,
            'ami',
            edition !== common_1.AmazonLinuxEdition.STANDARD ? edition : undefined,
            props.kernel?.toString(),
            props.virtualization ?? common_1.AmazonLinuxVirt.HVM,
            props.cpuType ?? common_1.AmazonLinuxCpuType.X86_64,
            props.storage ?? common_1.AmazonLinuxStorage.GENERAL_PURPOSE,
        ].filter(x => !!x);
        return '/aws/service/ami-amazon-linux-latest/' + parts.join('-');
    }
    constructor(props) {
        super({
            parameterName: AmazonLinux2ImageSsmParameter.ssmParameterName(props),
            cachedInContext: props.cachedInContext,
            userData: props.userData,
        });
        try {
            jsiiDeprecationWarnings.aws_cdk_lib_aws_ec2_AmazonLinux2ImageSsmParameterProps(props);
        }
        catch (error) {
            if (process.env.JSII_DEBUG !== "1" && error.name === "DeprecationError") {
                Error.captureStackTrace(error, AmazonLinux2ImageSsmParameter);
            }
            throw error;
        }
    }
}
_b = JSII_RTTI_SYMBOL_1;
AmazonLinux2ImageSsmParameter[_b] = { fqn: "aws-cdk-lib.aws_ec2.AmazonLinux2ImageSsmParameter", version: "2.76.0" };
exports.AmazonLinux2ImageSsmParameter = AmazonLinux2ImageSsmParameter;
